src/main/integrations/electron-breadcrumbs.ts

```diff
@@ -0,0 +1,206 @@
+import { addBreadcrumb, captureMessage, getCurrentHub } from '@sentry/core';
+import { NodeClient } from '@sentry/node';
+import { Breadcrumb, Integration } from '@sentry/types';
+import { app, autoUpdater, BrowserWindow, powerMonitor, screen, WebContents } from 'electron';
+
+import { whenAppReady } from '../electron-normalize';
+import { ElectronMainOptions } from '../sdk';
+
+/** A function that returns true if the named event should create breadcrumbs */
+type EventFunction = (name: string) => boolean;
+type EventTypes = boolean | string[] | EventFunction | undefined;
+
+interface ElectronBreadcrumbsOptions<EventTypes> {
+  /**
+   * Whether webContents `unresponsive` events are captured as Sentry events
+   *
+   * default: true
+   */
+  unresponsive: boolean;
+  /**
+   * app events
+   *
+   * default: (name) => !name.startsWith('remote-')
+   */
+  app: EventTypes;
+  /**
+   * autoUpdater events
+   *
+   * default: all
+   */
+  autoUpdater: EventTypes;
+  /**
+   * webContents events
+   * default: ['dom-ready', 'context-menu', 'load-url', 'destroyed']
+   */
+  webContents: EventTypes;
+  /**
+   * BrowserWindow events
+   *
+   * default: ['closed', 'close', 'unresponsive', 'responsive', 'show', 'blur', 'focus', 'hide',
+   *            'maximize', 'minimize', 'restore', 'enter-full-screen', 'leave-full-screen' ]
+   */
+  browserWindow: EventTypes;
+  /**
+   * screen events
+   *
+   * default: all
+   */
+  screen: EventTypes;
+  /**
+   * powerMonitor events
+   *
+   * default: all
+   */
+  powerMonitor: EventTypes;
+}
+
+const DEFAULT_OPTIONS: ElectronBreadcrumbsOptions<EventFunction> = {
+  unresponsive: true,
+  // We exclude events starting with remote as they can be quite verbose
+  app: (name) => !name.startsWith('remote-'),
+  autoUpdater: () => true,
+  webContents: (name) => ['dom-ready', 'context-menu', 'load-url', 'destroyed'].includes(name),
+  browserWindow: (name) =>
+    [
+      'closed',
+      'close',
+      'unresponsive',
+      'responsive',
+      'show',
+      'blur',
+      'focus',
+      'hide',
+      'maximize',
+      'minimize',
+      'restore',
+      'enter-full-screen',
+      'leave-full-screen',
+    ].includes(name),
+  screen: () => true,
+  powerMonitor: () => true,
+};
+
+/** Adds breadcrumbs for Electron events. */
+export class ElectronBreadcrumbs implements Integration {
+  /** @inheritDoc */
+  public static id: string = 'ElectronBreadcrumbs';
+
+  /** @inheritDoc */
+  public name: string = ElectronBreadcrumbs.id;
+
+  private readonly _options: ElectronBreadcrumbsOptions<EventFunction | undefined>;
+
+  /**
+   * @param _options Integration options
+   */
+  public constructor(options: Partial<ElectronBreadcrumbsOptions<EventTypes>> = {}) {
+    // Convert all 'true' properties to undefined since undefined uses the defaults too
+    (Object.keys(options) as (keyof ElectronBreadcrumbsOptions<EventTypes>)[]).map((k) => {
```

Comment:
We shouldn't use `.map` if we aren't gonna use the return value, `.forEach` probably applies better here.

Mutating the arg is not ideal though.

Think there's no way around the casting of the key due to: https://github.com/microsoft/TypeScript/pull/12253#issuecomment-263132208. `Object.keys` will probably never go generic :/

Reply:
Opened https://github.com/getsentry/sentry-javascript/pull/4229 to lint against the `.map` stuff.

Reply:
ah, and this doesn't handle `string[]` inputs either